hypothesis-python/src/hypothesis/provisional.py

```diff
@@ -51,17 +69,46 @@ def url_encode(s):
 
 
 @st.defines_strategy_with_reusable_values
-def domains():
-    """A strategy for :rfc:`1035` fully qualified domain names."""
-    atoms = st.text(
-        string.ascii_letters + "0123456789-", min_size=1, max_size=63
-    ).filter(lambda s: "-" not in s[0] + s[-1])
+def domains(
+    max_length=255,  # type: int
+    max_element_length=63,  # type: int
+):
+    # type: (...) -> SearchStrategy[Text]
+    """A strategy for :rfc:`1035` fully qualified domain names.
+
+    The max_length is 255 in accordance with :rfc:`1035#section-2.3.4`
+    The maximum_element_length is 63 in accordance with :rfc:`1035#section-2.3.4`
+    """
+    # https://tools.ietf.org/html/rfc1035#section-2.3.4
+    if max_length > 255:
+        raise InvalidArgument(
+            u"Invalid max_length %s. max_length must be shorter than 255.", max_length
+        )
+
+    if max_element_length > 63:
+        raise InvalidArgument(
+            u"Invalid max_element_length %s. max_element_length must be less than 63.",
+            max_element_length,
+        )
```

Comment:
Yeah, i should use the `check_valid_size` to catch the negative values. I initially dismissed it because I wanted a way to catch the maximums permissible values and it didn't do that, also if I start using `check_valid_size` it makes sense to try and have similar style output. I'll clean this up.

Reply:
Note that `check_valid_size` allows `None`, so you'll also need `check_type`.

(it's a lot of validation, but that's what it takes to guard against misuse...)

Reply:
@Zac-HD `check_valid_size` actually rejects none. It was deprecated. 😄 https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/src/hypothesis/internal/validation.py#L107

Reply:
Only for the specific name "min_size" though, and we're dealing with length params.

Reply:
I've replaced this with a new check_function function. It seemed weird to have functions to check types, maximums, and that ranges are specified correctly (low < high) but nothing for checking min < value < max.